cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KRaftVersionChangeCreator.java

```diff
@@ -0,0 +1,276 @@
+     *
+     * @return  Future which completes when the Kafka broker pods are retrieved
+     */
+    private Future<List<Pod>> getPods()   {
+        Labels selectorLabels = Labels.forStrimziKind(Kafka.RESOURCE_KIND)
+                .withStrimziCluster(reconciliation.name())
+                .withStrimziName(KafkaResources.kafkaStatefulSetName(reconciliation.name()));
+
+        return podOperator.listAsync(reconciliation.namespace(), selectorLabels);
+    }
+
+    /**
+     * Detects the current and desired Kafka versions based on the information collected from Kubernetes
+     *
+     * @return  Future which completes when the "to" and "from" versions are collected
+     */
+    private Future<Void> detectToAndFromVersions(List<Pod> pods)    {
+        String lowestKafkaVersion = null;
+        String highestKafkaVersion = null;
+
+        for (Pod pod : pods)    {
+            // Collect the Kafka version from the pods
+            String currentVersion = Annotations.stringAnnotation(pod, ANNO_STRIMZI_IO_KAFKA_VERSION, null);
+
+            // We find the highest and lowest used Kafka version. This is used to detect any upgrades or
+            // downgrades which failed in the middle and continue with them.
+            if (currentVersion != null) {
+                if (highestKafkaVersion == null)  {
+                    highestKafkaVersion = currentVersion;
+                } else if (compareDottedVersions(highestKafkaVersion, currentVersion) < 0) {
+                    highestKafkaVersion = currentVersion;
+                }
+
+                if (lowestKafkaVersion == null)  {
+                    lowestKafkaVersion = currentVersion;
+                } else if (compareDottedVersions(lowestKafkaVersion, currentVersion) > 0) {
+                    lowestKafkaVersion = currentVersion;
+                }
+            }
+        }
+
+        // We decide what is the current Kafka version used and create the KafkaVersionChange object
+        // describing the situation.
+        if (lowestKafkaVersion == null)   {
+            if (pods.isEmpty())  {
+                // No pods were found. We use the desired version from the Kafka CR
+                versionFrom = versionFromCr;
+                versionTo = versionFromCr;
+            } else {
+                // Pods already exist. However, none of them contains the version annotation. This suggests they are not
+                // created by any recent versions of Strimzi. Without the annotation, we cannot detect the Kafka version
+                // and decide on upgrade.
+                LOGGER.warnCr(reconciliation, "Kafka Pods exist, but do not contain the {} annotation to detect their version. Kafka upgrade cannot be detected.", ANNO_STRIMZI_IO_KAFKA_VERSION);
+                throw new KafkaUpgradeException("Kafka Pods exist, but do not contain the " + ANNO_STRIMZI_IO_KAFKA_VERSION + " annotation to detect their version. Kafka upgrade cannot be detected.");
```

Comment:
What exactly you mean by duplication?

Reply:
Some warning message strings are duplicated, you can build it just once and reuse.

Reply:
I think that is how we do it in general. You don't really get much from creating one String and reusing it for a message like this. Just more complex code.

Reply:
> I think this is how we do it in general. You don't really get much from creating one String and reusing it for a message like this. Just more complex code.

I think this is wrong, especially if we do it in general. As you can see, you have two in-memory objects instead of one, for each error message.

![Screenshot from 2023-11-29 10-37-36](https://github.com/strimzi/strimzi-kafka-operator/assets/11456498/4f9a0b0d-5db6-4f8b-bd96-ad57a7a04087)

Additionally, I don't see how the following code is more complex than the original.

```sh
String message = "Kafka Pods exist, but do not contain the " + ANNO_STRIMZI_IO_KAFKA_VERSION + " annotation to detect their version. Kafka upgrade cannot be detected."
LOGGER.warnCr(reconciliation, message, ANNO_STRIMZI_IO_KAFKA_VERSION);
throw new KafkaUpgradeException(message);
```

Reply:
It is also true that log4j builds the string only when logged, so in the end it's not a big issue, but I would prefer to avoid the duplication.

Reply:
I prefer to keep it as it is. Sorry.